xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiCacheStore.java

```diff
@@ -400,6 +400,15 @@ public XWikiDocument loadXWikiDoc(XWikiDocument doc, XWikiContext inputxcontext)
 
                     LOGGER.debug("Document [{}] was retrieved from persistent storage", key);
 
+                    // If the exists cache indicated that the document should exist, log an error with a
+                    // stacktrace.
+                    if (cachedoc.isNew() && result != null) {
+                        LOGGER.error("Document [{}] was not found in persistent storage but the exists cache "
+                                + "indicated that it should exist. Unless that document has just been deleted, this "
+                                + "is a bug and should be reported.", key,
```

Comment:
Not sure key is best here to identify the doc by humans, I might have used the ref as it's more human readable.

Reply:
Also I'd probably have put in the message that this log is for identifying the root cause of XWIKI-21331: it might help if some users see those in their console.